src/app/games/quarto/QuartoRules.ts

```diff
@@ -129,41 +129,175 @@ interface LineInfos {
     boardStatus: MGPOptional<BoardStatus>;
 }
 
-export class QuartoRules extends Rules<QuartoMove, QuartoState> {
+abstract class VictoryPattern {
+
+    protected constructor(public readonly level: number,
+                          public readonly coordPattern: CoordSet,
+                          public readonly initialCoord: Coord,
+                          public readonly remaining: number,
+    ) {}
+
+    public abstract getNextPattern(): VictoryPattern;
+
+    public getCoords(): Set<Coord> {
+        return this.coordPattern.map((element: Coord) => element.getNext(this.initialCoord));
+    }
+}
+
+class VerticalVictoryPattern extends VictoryPattern {
+
+    private static readonly VERTICAL: CoordSet = new CoordSet([
+        new Coord(0, 0),
+        new Coord(0, 1),
+        new Coord(0, 2),
+        new Coord(0, 3),
+    ]);
+
+    public constructor(initialCoord: Coord, remaining: number) {
+        super(1,
+              VerticalVictoryPattern.VERTICAL,
+              initialCoord,
+              remaining);
+    }
+
+    public override getNextPattern(): VerticalVictoryPattern {
+        const newCoord: Coord = this.initialCoord.getNext(Orthogonal.RIGHT, 1);
+        const newRemaining: number = this.remaining - 1;
+        return new VerticalVictoryPattern(newCoord, newRemaining);
+    }
+
+}
+
+class HorizontalVictoryPattern extends VictoryPattern {
+
+    private static readonly HORIZONTAL: CoordSet = new CoordSet([
+        new Coord(0, 0),
+        new Coord(1, 0),
+        new Coord(2, 0),
+        new Coord(3, 0),
+    ]);
+
+    public constructor(initialCoord: Coord, remaining: number) {
+        super(1,
+              HorizontalVictoryPattern.HORIZONTAL,
+              initialCoord,
+              remaining);
+    }
+
+    public override getNextPattern(): HorizontalVictoryPattern {
+        const newCoord: Coord = this.initialCoord.getNext(Orthogonal.DOWN, 1);
+        const newRemaining: number = this.remaining - 1;
+        return new HorizontalVictoryPattern(newCoord, newRemaining);
+    }
+
+}
+
+class DescendingDiagonalVictoryPattern extends VictoryPattern {
+
+    private static readonly DESCENDING_DIAGONAL: CoordSet = new CoordSet([
+        new Coord(0, 0),
+        new Coord(1, 1),
+        new Coord(2, 2),
+        new Coord(3, 3),
+    ]);
+
+    public constructor(initialCoord: Coord, remaining: number) {
+        super(1,
+              DescendingDiagonalVictoryPattern.DESCENDING_DIAGONAL,
+              initialCoord,
+              remaining);
+    }
+
+    public override getNextPattern(): DescendingDiagonalVictoryPattern {
+        Utils.assert(false, 'there is only one descending diagonal on normal quarto board');
```

Comment:
assert false should be throw
et puis c'est une indication de mauvais code ici je dirais non ?

Reply:
ouais ça aurais du passer par deux sous classe "singleton victory pattern" et "multiple victory pattern" et celle ci aurais été singleton mais on a changé de façon de faire donc good good